### DISCUSSION TITLE
TypeError: Cannot read properties of undefined (reading 'db')

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#37523] [Help] [answered]
So I have this portion of code that goes like this:
from this page here:
https://github.com/Elindo586/nextprojectb/blob/main/pages/inventory.js

```
import { clientPromise } from "../utils/mongodb";

export async function getStaticProps() {
  // await clientPromise;
  // `await clientPromise` will use the default database passed in the MONGODB_URI
  // However you can use another database (e.g. myDatabase) by replacing the `await clientPromise` with the following code:
  //
  const client = await clientPromise;
  const dbs = client.db("parts");
  //
  // Then you can execute queries against your database like so:
  // db.find({}) or any of the MongoDB Node Driver commands

  return {
    props: { db },
    revalidate: 60,
  };
}

const Inventory = ({ db }) => {
```

I am getting an error in line 15 like this:
```
error - pages\inventory.js (15:21) @ getStaticProps
TypeError: Cannot read properties of undefined (reading 'db')
  13 |   //
  14 |   const client = await clientPromise;
> 15 |   const db = client.db("parts");
     |                     ^
```
I am importing the connection from utils/mongodb here..
https://github.com/Elindo586/nextprojectb/blob/main/utils/mongodb.js

^^^ this connection I think it is working as I am not getting any errors there any more.

But in the page in which I am calling for the connection, and then trying to extract the collection doesn't seem to recognize  the part of  `client.db`

**Comment by Elindo586** [comment]
```
 at getStaticProps (webpack-internal:///./pages/inventory.js:30:23)
    at async Object.renderToHTML (file://C:\Users\Edgar\Desktop\Stuff\Acode\nextprojectb\node_modules\next\dist\server\render.js:492:20)
    at async doRender (file://C:\Users\Edgar\Desktop\Stuff\Acode\nextprojectb\node_modules\next\dist\server\base-server.js:901:38)
    at async cacheEntry.responseCache.get.isManualRevalidate.isManualRevalidate (file://C:\Users\Edgar\Desktop\Stuff\Acode\nextprojectb\node_modules\next\dist\server\base-server.js:1006:28)
    at async (file://C:\Users\Edgar\Desktop\Stuff\Acode\nextprojectb\node_modules\next\dist\server\response-cache.js:74:36)
```

**Comment by icyJoseph** [comment]
You await the connection, and once that resolves, you gotta fetch the schema.

```ts
import { clientPromise } from "../utils/mongodb";

export async function getStaticProps() {
  await clientPromise; 
  // now you need to create a Model to query from
  const parts = mongoose.model('Parts', yourSchema);
  const props = parts.find({}) // not sure about this call

  return {
    props, // props must be serializable!
    revalidate: 60,
  };
}
```

Please read the documentation and example with mongoDB and mongoose. Here's an example of an schema, https://github.com/vercel/next.js/blob/canary/examples/with-mongodb-mongoose/models/Pet.js.

**Comment by Elindo586** [comment]
I think it is only fair to share my feedback frustration on this connection.

1.	I am just a part-part-part time online learner for coding.  I might not have all the knowledge or smart enough to figure it out.
2.	During COVID times, family life, and work, my coding time is very limited 
3.	I have tried lots of samples, the last sample I am trying is from this Vercel link here: https://github.com/vercel/next.js/tree/canary/examples/with-mongodb
4.	Given that I wasn’t able to extract the data, I went ahead and paid $100 for a fiverr guy who claims to work with Nextjs, and after a week, they couldn’t extract the data either…  yes, I got my money back.
5.	Given the Nextjs docs, the data is usually extracted from a live external link, and not usually the typical MERN way to do it with MongoDB or Mongoose and REACT..
6.	Given that we are talking about MongoDB, at times there might just be one silly thing to do and fix the issue… (like if I want to upload and excel file to their database I have to run the command from the location in which I installed Mongosh)… 
7.	I can also try the API key/value link and see if I can extract that data that way, but given that I have tried for over 2 weeks and that I am talking about some silly things that I might not know with MongoDB.. then I’ll skip that try..
8.	There was an article I was reading/trying hosted in MongoDB about this exact operation… after I asked more details on the article, the article was later changed..
9.	Like I said… I don’t have all the smarts to figure it out and on the other hand I have to weight that it could very well be something silly from MongoDB 

Next step is to try the data coming from a local json file…….  Given my part-part-part time coding time, hopefully I can do it this week, or by the end of next week if I squeeze the time to try it.

**Reply by icyJoseph** [reply]
Good to be open and vent out about your experience. None of these concepts are easy, and it takes time to get a good grip. 

It is a bit of a non-starter, but it is important try and not get frustrated at these things, at least not to a personal level. These API's, examples and documentation are all made with tons of assumptions about knowledge and experience (even though the authors try to keep it clean, a lot of bias leaks in), and initially we all feel the burden of having to learn, and even re-learn, these things. I for once, have avoided working with MongoDB as much as I could. 

I think by trying out different approaches and formulating solutions, you are growing learning more than you realise, but perhaps tackling it all in one project is a bit much. Unfortunately, setting up a dummy project with mongo db to try and work out the API details is not that simple either, and it will take time too. 

I think in particular point 5 is where things might be colliding for you, because the only difference is that the data can be fetched inside a function that then injects the results as props to a page. And that has its own benefits and gains over a Client Side Rendered application.

Though I cannot help much more, this is good information, it means that the documentation on this regard is not clear enough, and that the example lacks more clarity. 

Have you tried looking for Next.js with MongoDB courses online? I don't want to recommend particular sites, or authors, but I know there's good content out there.

**Comment by Elindo586** [comment]
I have taken some REACT and Next.js courses....  in REACT usually you get some talk about MongoDB / Mongoose and how to connect it..

I think there are free courses from MongoDB as well.

A lot of the courses do talk about details of operations, either from REACT, Next, or MongoDB depending on what course you are taking.

Searching in Udemy for example, I don't think there is a course that will touch in the details on how to do a MongoDB connection to any of the getProps from Next......   this type of connection seems to be the standard in REACT, but not mentioned in Next....

Also... the Next documentation and most of connection videos I have found don't use MongoDB, instead the connection comes from a live link API which is different than using connection and schema from MongoDB or Mongoose..   MongoDB can also provide an API link with a key, similar to the data extraction done on the Next.js samples, but  that is another can of worms that I don't feel like trying by now. 

If I can get this working from a json file that would do...

In the future I am however more interested in learning about API's so I can manage my own data...

**Reply by icyJoseph** [reply]
That's good attitude, what do you think of this video? https://www.youtube.com/watch?v=cM0pA50R20M, from minute 5 to 7 he describes your problem very well, and explains how to implement it. 

He then goes on to use the Test model in api folders, but you could use it inside getStaticProps too. Then at around minute 20 he uses the same Test model on getServerSideProps, which in this case, is pretty much the same as using it in getStaticProps. He also explains why we have to do `JSON.parse(JSON.stringify(data))` trick!

I guess in your case, it should be called `Parts`?

**Comment by Elindo586** [comment]
That is interesting... I am gonna give it a try when I get some time.    He is using Mongoose for the connection instead of MongoDB...

**Comment by Elindo586** [comment, marked as answer]
Mother of God...

I got the stuff working from the json file.... 

I think the last sample video was cool, but the json will be sufficient for now.

In the future I want to dive into API's  so that I get to have a lot more fun with them...


Next step is working on the deployment.... I got some errors after running npm build from my computer... I'll check that later...      hopefully the deployment is no way as hard as getting the Mongo data..

**Reply by asthaarora95** [reply]
Hi I am facing the same error, can you tell me in detail what did you do to resolve it?? asap
